### PR TITLE
refactor(switch): use composition api

### PR DESCRIPTION
**Pull request by ajuner**
First of all, thank you for your contribution! 😄

New feature please send pull request to feature branch, and rest to master branch. Pull request will be merged after one of collaborators approve. Please makes sure that these form are filled before submitting your pull request, thank you!

[[中文版模板 / Chinese template](https://github.com/vueComponent/ant-design-vue/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [x] Refactoring
- [ ] Code style optimization
- [ ] Branch merge
- [ ] Other (about what?)

### What's the background?

> 1. Describe the source of requirement.
> 2. Resolve what problem.
> 3. Related issue link.

### API Realization (Optional if not new feature)

> 1. Basic thought of solution and other optional proposal.
> 2. List final API realization and usage sample.
> 3. GIF or snapshot should be provided if includes UI/interactive modification.

### What's the effect? (Optional if not new feature)

> 1. Does this PR affect user? Which part will be affected?
> 2. What will say in changelog?
> 3. Does this PR contains potential break change or other risk?

### Changelog description (Optional if not new feature)

> 1. English description
> 2. Chinese description (optional)

### Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### Additional Plan? (Optional if not new feature)

> If this PR related with other PR or following info. You can type here.

### CODE REVIEW

**Comment by ajuner** [comment]
大佬们帮忙看看

**Comment by ajuner** [comment]
还有要不要把defaultChecked去掉好了

**Comment by tangjinzhou** [comment]
如果去掉就是破坏性更新  default 系列在 3.0 里面移除吧  现在给即将废弃的  warning 提示

**Comment by ajuner** [comment]
> 如果去掉就是破坏性更新 default 系列在 3.0 里面移除吧 现在给即将废弃的 warning 提示

Ok

**Comment by ajuner** [comment]
我改改

File: components/switch/index.tsx [hunk 1]
```diff
@@ -1,86 +1,85 @@
...
 import PropTypes from '../_util/vue-types';
-import hasProp, { getOptionProps, getComponent } from '../_util/props-util';
 import VcSwitch from '../vc-switch';
 import Wave from '../_util/wave';
 import { defaultConfigProvider } from '../config-provider';
 import warning from '../_util/warning';
 import { tuple, withInstall } from '../_util/type';
+import Omit from 'omit.js';
+
+function getDefaultSlot(slots: Record<string, any>, props: Record<string, any>, prop: string) {
```
Comment (Amour1688) [inline components/switch/index.tsx:11]: 这个在 util 里面应该有

File: components/switch/index.tsx [hunk 2]
```diff
@@ -1,86 +1,85 @@
...
+      }
+      warning(
+        'checked' in props || !('value' in attrs),
+        'Switch',
+        '`value` is not validate prop, do you mean `checked`?',
+      );
+    });
+    const { getPrefixCls } = configProvider;
+
+    const prefixCls = getPrefixCls('switch', props.prefixCls);
```
Comment (Amour1688) [inline components/switch/index.tsx:62]: 这样写 props 里面的 prefixCls 更新了，这个 prefixCls 不会更新

File: components/switch/index.tsx [hunk 3]
```diff
@@ -1,86 +1,85 @@
...
-      this.refSwitchNode?.focus();
-    },
-    blur() {
-      this.refSwitchNode?.blur();
-    },
-    saveRef(c) {
-      this.refSwitchNode = c;
-    },
-  },
+  props: switchProps(),
```
Comment (Amour1688) [inline components/switch/index.tsx:36]: switchProps 没有在其他地方用到，需要搞成一个函数吗

Comment (ajuner) [reply components/switch/index.tsx:36]: 为了做上面SwitchProps类型导出，有什么更好点的方法嘛

File: components/switch/index.tsx (hunk 1 above)
Comment (ajuner) [reply components/switch/index.tsx:11]: 我再去仔细翻一遍

Comment (ajuner) [reply components/switch/index.tsx:11]: 在prop-util下只找到了之前OptionsAPI的用法getComponent，要不把这个方法放到里面去

Comment (zkwolf) [reply components/switch/index.tsx:11]: 放util吧，我记得之前你好像还写过一个一模一样的，可能在其他组件里

Comment (ajuner) [reply components/switch/index.tsx]: 嗯嗯  等这个合了我把其他地方也改下

File: components/switch/index.tsx [hunk 4]
```diff
@@ -1,86 +1,84 @@
...
+  setup(props: Record<string, any>, { attrs, slots, expose }) {
```
Comment (Amour1688) [inline components/switch/index.tsx:34]: 不应该是 Record

File: components/switch/index.tsx (hunk 3 above)
Comment (Amour1688) [reply components/switch/index.tsx:33]: 只是为了类型导出不需要函数

File: components/vc-switch/src/Switch.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
+import { switchPropTypes } from './PropTypes';
+import Omit from 'omit.js';
+import { defineComponent, nextTick, onMounted, ref } from 'vue';
+import KeyCode from '../../_util/KeyCode';
+
+function getComponent(slots: Record<string, any>, props: Record<string, any>, prop: string) {
```
Comment (Amour1688) [inline components/vc-switch/src/Switch.tsx:6]: 可以用 util 里的

**Comment by ajuner** [comment]
done

File: components/switch/__tests__/__snapshots__/index.test.js.snap [hunk 6]
```diff
@@ -1,7 +1,7 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
 exports[`Switch should has click wave effect 1`] = `
-<button class="ant-switch ant-switch-checked" type="button" role="switch" aria-checked="true">
+<button prefixcls="ant-switch" class="ant-switch ant-switch-checked" type="button" role="switch" aria-checked="true">
```
Comment (zkwolf) [inline components/switch/__tests__/__snapshots__/index.test.js.snap]: 这里不对啊，prefixCls不应该被加上来

Comment (ajuner) [reply components/switch/__tests__/__snapshots__/index.test.js.snap]: 对 我改了

Comment (ajuner) [reply components/switch/__tests__/__snapshots__/index.test.js.snap]: 我再检查下  rebase后好像出了点问题

File: components/switch/index.tsx [hunk 7]
```diff
@@ -1,86 +1,84 @@
...
+    });
+    const { getPrefixCls } = configProvider;
+    return () => (
       <Wave insertExtraNode>
-        <VcSwitch {...switchProps} />
+        <VcSwitch
+          {...Omit(props, ['prefixCls', 'size', 'loading', 'disabled'])}
+          {...attrs}
+          checked={props.checked}
+          prefixCls={getPrefixCls('switch', props.prefixCls)}
```
Comment (Amour1688) [inline components/switch/index.tsx:66]: 抽个变量出来吧

File: components/vc-switch/src/Switch.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import { getPropsSlot } from '../../_util/props-util';
+
+export default defineComponent({
+  name: 'VcSwitch',
+  inheritAttrs: false,
+  props: {
+    ...switchPropTypes,
+    prefixCls: switchPropTypes.prefixCls.def('rc-switch'),
+  },
+  setup(props, { attrs, slots, emit, expose }) {
```
Comment (Amour1688) [inline components/vc-switch/src/Switch.tsx:14]: emits 不用声明吗

Comment (ajuner) [reply components/vc-switch/src/Switch.tsx:15]: 还是声明一下好

File: components/vc-switch/src/Switch.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+export default defineComponent({
+  name: 'VcSwitch',
+  inheritAttrs: false,
+  props: {
+    ...switchPropTypes,
+    prefixCls: switchPropTypes.prefixCls.def('rc-switch'),
+  },
+  emits: ['update:checked', 'mouseup', 'change', 'click'],
+  setup(props, { attrs, slots, emit, expose }) {
+    const checked = ref('checked' in props ? !!props.checked : !!props.defaultChecked);
```
Comment (tangjinzhou) [inline components/vc-switch/src/Switch.tsx:16]: 没有 watch checked

File: components/vc-switch/src/Switch.tsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        }
+      });
+    });
+
+    const setChecked = (check: boolean, e: MouseEvent | KeyboardEvent) => {
+      if (props.disabled) {
+        return;
+      }
+      checked.value = !checked.value;
+      emit('update:checked', checked);
```
Comment (tangjinzhou) [inline components/vc-switch/src/Switch.tsx:33]: emit('update:checked', checked.value);

**Pull request merged**